### PR TITLE
feat(*): install remote and local plugins

### PR DESCRIPTION
**Pull request by michelleN**
This implements the plugin install command. The bulk of this
creates a plugin installer specific for the Draft environment and
it creates the plugin install command with the proper setup.
Addresses #185 by creating a framework for a plugin manager.

### CODE REVIEW

**Comment by bacongobbler** [comment]
Looks like CI is borking on the missing package. Would you mind adding github.com/Masterminds/vcs to glide.yaml and run `glide up`?

**Comment by michelleN** [comment]
did we disable CI @bacongobbler ? I'm not seeing any builds now.

**Comment by bacongobbler** [comment]
Nah it was probably just backlogged. There is only one node available to run CI builds and we share that single node with Deis Workflow. It's one of the reasons why I really wanted to get off of https://ci.deis.io ASAP, but that's another story for another day.

It looks like they're running now. :)

File: cmd/draft/plugin.go [hunk 1]
```diff
@@ -125,24 +161,37 @@ func manuallyProcessArgs(args []string) ([]string, []string) {
...
+func runHook(p *plugin.Plugin, event string) error {
+	hook := p.Metadata.Hooks.Get(event)
+	if hook == "" {
+		return nil
+	}
+
+	prog := exec.Command("sh", "-c", hook)
+	// TODO make this work on windows
+	// I think its ... ¯\_(ツ)_/¯
+	// prog := exec.Command("cmd", "/C", p.Metadata.Hooks.Install())
```
Comment (bacongobbler) [inline cmd/draft/plugin.go]: We had to do something similar for draft packs. You could take advantage of that!

https://github.com/Azure/draft/blob/master/pkg/draft/pack/pack.go#L84-L90

File: cmd/draft/plugin_list.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	out  io.Writer
+}
+
+func newPluginListCmd(out io.Writer) *cobra.Command {
+	pcmd := &pluginListCmd{out: out}
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "list installed Draft plugins",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			//TODO: check on draftHome() to make sure no funky stuff with envs is happening...
```
Comment (bacongobbler) [inline cmd/draft/plugin_list.go]: Is this something that still needs to be done? If so, can you clarify what is the problem at hand? :)

File: pkg/plugin/installer/base.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	"path/filepath"
+
+	"github.com/Azure/draft/pkg/draft/draftpath"
+)
+
+type base struct {
+	// Source is the reference to a plugin
+	Source string
+
+	//TODO: change DraftHome to ProjectHome or something more generic
```
Comment (bacongobbler) [inline pkg/plugin/installer/base.go]: does this still need to be done? I'm okay with whichever makes more sense here.

File: pkg/plugin/installer/installer.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+func debug(format string, args ...interface{}) {
+	if Debug {
+		format = fmt.Sprintf("[debug] %s\n", format)
+		fmt.Printf(format, args...)
+	}
+}
+
+// isLocalReference checks if the source exists on the filesystem.
+func isLocalReference(source string) bool {
+	//TODO: can user install a local plugin using "."?
```
Comment (bacongobbler) [inline pkg/plugin/installer/installer.go]: I would say no, it has to be a path for now (even if the path is just `./`. We should document that.

File: cmd/draft/plugin.go (hunk 1 above)
Comment (bacongobbler) [inline cmd/draft/plugin.go]: I can see if I can round up a Windows VM to test future PRs but I wouldn't fret too much. We can iron out the Windows kinks once we get someone or something regularly testing this on Windows.

File: cmd/draft/plugin_list.go (hunk 2 above)
Comment (michelleN) [inline cmd/draft/plugin_list.go]: oh good catch. I actually already addressed this in the run cmd with the pluginDirPath() func. will delete this comment.

File: pkg/plugin/installer/installer.go (hunk 4 above)
Comment (michelleN) [inline pkg/plugin/installer/installer.go]: deleting this as well

File: cmd/draft/plugin.go (hunk 1 above)
Comment (michelleN) [inline cmd/draft/plugin.go]: sounds good

File: pkg/plugin/installer/base.go (hunk 3 above)
Comment (michelleN) [inline pkg/plugin/installer/base.go]: nope. deleting.

**Comment by michelleN** [comment]
@bacongobbler addressed all the comments and added documentation.

File: docs/plugins-guide.md [hunk 5]
```diff
@@ -18,23 +19,24 @@ Draft plugins have the following features:
...
-while the plugins do the "detail work" of performing a desired action.
+The Draft plugin model is partially modeled on Git's plugin model. To that end, you may sometimes hear `draft` referred to as the _porcelain_ layer, with plugins being the _plumbing_. This is a shorthand way of suggesting that Draft provides the user experience and top level processing logic, while the plugins do the "detail work" of performing a desired action.
 
 ## Installing a Plugin
 
-A Draft plugin management system is in the works. But in the short term, plugins are installed by
-copying the plugin directory into `$(draft home)/plugins`.
+Plugins are installed using `$draft plugin install <plugin url/path>`. You can pass in a url to a git repository which contains the plugin or you can opt to pass in a local path to a plugin on your machine.
+```console
+$ draft plugin install https://github.com/michelleN/draft-server
```
Comment (bacongobbler) [inline docs/plugins-guide.md]: 👍 nice plugin.

**Comment by bacongobbler** [comment]
merge when ready!

**Comment by bacongobbler** [comment]
Works as advertised.

```
><> draft plugin install https://github.com/michelleN/draft-server
Installed plugin: server
><> draft plugin list
NAME  	VERSION	DESCRIPTION                               
server	0.1.0  	Helpers for working with in-cluster draftd
><> draft server
Helpers for working with in-cluster tiller.

Available Commands:
  destroy       Delete draftd from cluster

Error: plugin "server" exited with error
```

🎉

**Pull request merged**